### PR TITLE
Fix disappearing labels of reference attributes

### LINKED ISSUE #4219: [Bug]: Assign reference button not aligned
### Description of the issue

When creating a product reference button that assigns reference is not aligned

### Steps to reproduce the problem

1. Create an attribute that enables to add references 
2. Create prod type with attribute from 1st step
3. Create a product and add the attribute from the previous step
4. Navigate to Catalog/Products and Create a new product

Assign button misplaced:

Zrzut ekranu 2023-09-21 o 11 12 50

Assignments misplaced as well (id instead of name is reported [here](https://github.com/saleor/saleor-dashboard/issues/2861)): 

Zrzut ekranu 2023-09-21 o 11 17 43

### What did you expect to happen?

Button aligned according to rest of the page

### Additional information

_No response_

### Environment

Dashboard version: 3.16

### PR DESCRIPTION
**Pull request by Droniu**
Closes #4219, #2861, #4272.

### Acceptance Criteria:
- [ ] As a user, I should be able to change product / page / variant reference attribute on product create view,
- [ ] As a user, I should be able to change product / page / variant reference attribute on product details view,
- [ ] As a user, I should be able to change product / page / variant reference attribute on page create view,
- [ ] As a user, I should be able to change product / page / variant reference attribute on page details view,
- [ ] As a user, I should be able to change product / page / variant reference attribute on variant create view,
- [ ] As a user, I should be able to change product / page / variant reference attribute on variant details view,
- [ ] As a user, I should be able to change eligible entities - category / collection / product / variant on sale details view,
- [ ] As a user, I should be able to change eligible entities - category / collection / product on voucher details view,

### Pull Request Checklist

1. [x] This code contains UI changes
2. [ ] All visible strings are translated with proper context including data-formatting
3. [ ] Attributes `data-test-id` are added for new elements
4. [ ] The changes are tested in Chrome/Firefox/Safari browsers and in light/dark mode
5. [x] Your code works with the latest stable version of the core
6. [x] I added changesets file (instructions in [contribution guide](./CONTRIBUTING.md)

### Test environment config

API_URI=https://automation-dashboard.staging.saleor.cloud/graphql/
APPS_MARKETPLACE_API_URI=https://apps.staging.saleor.io/api/v2/saleor-apps

### Do you want to run more stable tests?

To run all tests, just select the stable checkbox. To speed up tests, increase the number of containers. Tests will be re-run only when the "run e2e" label is added.

1. [x] stable
2. [ ] app
3. [x] attribute
4. [ ] category
5. [ ] collection
6. [ ] customer
7. [ ] giftCard
8. [ ] homePage
9. [ ] login
10. [ ] menuNavigation
11. [ ] navigation
12. [ ] orders
13. [x] pages
14. [ ] payments
15. [ ] permissions
16. [ ] plugins
17. [ ] productType
18. [x] products
19. [x] sales
20. [ ] shipping
21. [ ] translations
22. [x] variants
23. [x] vouchers

CONTAINERS=4

### CODE REVIEW

File: src/hooks/makeTopLevelSearch/makeTopLevelSearch.ts [hunk 1]
```diff
@@ -4,10 +4,10 @@ import { DocumentNode } from "graphql";
 
 import makeSearch, { SearchVariables, UseSearchHook } from "../makeSearch";
 
-export interface SearchData {
+export interface SearchData<T = any> {
```
Comment (poulch) [inline src/hooks/makeTopLevelSearch/makeTopLevelSearch.ts:7]: question: can you try to use unknow here?

File: src/hooks/useFormset.ts [hunk 2]
```diff
@@ -60,10 +66,25 @@ function useFormset<TData = {}, TValue = any>(
...
+    setData(data => {
+      const itemIndex = data.findIndex(item => item.id === id);
+      return [
+        ...data.slice(0, itemIndex),
+        {
+          ...data[itemIndex],
+          metadata,
+        },
+        ...data.slice(itemIndex + 1),
+      ];
```
Comment (poulch) [inline src/hooks/useFormset.ts:79]: I'd recommend use map or reduce here, current solution is not the best in terms of speed and clearness

```suggestion
     return data.map((item) => {
        if (item. id === id) {
            return {
                ...item,
                metadata
            }
        }
         return item
     })
```

File: src/components/Attributes/utils.ts [hunk 3]
```diff
@@ -76,6 +76,17 @@ export function getReferenceDisplayValue(
       return definedAttributeReference;
     }
 
+    // If value has not been yet assigned and data
+    // is no longer available, use metadata
+    if (attribute.metadata) {
+      return {
+        label: attribute.metadata.find(
+          metadata => metadata.value === attributeValue,
+        )?.label,
```
Comment (poulch) [inline src/components/Attributes/utils.ts:85]: question: why you keep metadata in array instead of an object?

File: src/hooks/useFormset.ts (hunk 2 above)
Comment (andrzejewsky) [reply src/hooks/useFormset.ts:79]: I would recommend to split this into function - I have no clue what's happening here, the name of the function can explain a bit

**Review by andrzejewsky — CHANGES_REQUESTED** [review]
Please attach changeset file

File: src/components/Attributes/utils.ts (hunk 3 above)
Comment (Droniu) [reply src/components/Attributes/utils.ts:85]: We could choose `key:value` as `id:label` pairs here but it's simpler to translate `Container[]` type returned from Assign Dialogs to array than single object

**Review by poulch — DISMISSED** [review]

**Review by andrzejewsky — APPROVED** [review]

**Comment by wojteknowacki** [comment]
Reference section disappears when saving product
https://github.com/saleor/saleor-dashboard/assets/124166231/85a2e1af-93ea-4dc2-9bc4-62f5d0a85a06

**Comment by wojteknowacki** [comment]
Can not assign reference on create content page

https://github.com/saleor/saleor-dashboard/assets/124166231/1a509793-eef8-4cb6-b985-a43efe77febe

**Comment by wojteknowacki** [comment]
The reference section disappears when saving variant

https://github.com/saleor/saleor-dashboard/assets/124166231/fee1bd68-d17e-4e75-acd7-4a4ba15e4ef3

File: src/components/SortableChip/SortableChip.tsx [hunk 4]
```diff
@@ -56,20 +60,23 @@ const SortableChip = SortableElement((props: SortableChipProps) => {
   };
 
   return (
-    <div className={clsx(classes.root, className)}>
+    <div className={clsx(classes.root, loading && classes.disabled, className)}>
```
Comment (poulch) [inline src/components/SortableChip/SortableChip.tsx:63]:
```suggestion
    <div className={clsx(classes.root, className, {
        [classes.disabled]: disabled
    )}>
```

File: src/components/SortableChip/SortableChip.tsx [hunk 5]
```diff
@@ -56,20 +60,23 @@ const SortableChip = SortableElement((props: SortableChipProps) => {
...
       <div className={classes.content}>
         <SortableHandle
-          className={classes.sortableHandle}
+          className={clsx(classes.sortableHandle, loading && classes.disabled)}
```
Comment (poulch) [inline src/components/SortableChip/SortableChip.tsx:66]:
```suggestion
          className={clsx(classes.sortableHandle, {
              [classes.disabled]: loading
          )}
```

File: src/components/SortableChip/SortableChip.tsx [hunk 6]
```diff
@@ -56,20 +60,23 @@ const SortableChip = SortableElement((props: SortableChipProps) => {
...
           data-test-id="button-drag-handle"
         />
         <Typography data-test-id="chip-label">{label}</Typography>
         {onClose && (
           <button
-            className={classes.closeButton}
+            className={clsx(classes.closeButton, loading && classes.disabled)}
```
Comment (poulch) [inline src/components/SortableChip/SortableChip.tsx:72]:
```suggestion
     className={clsx(classes.sortableHandle, {
              [classes.disabled]: loading
          )}
```

File: src/components/SortableChip/SortableChip.tsx [hunk 7]
```diff
@@ -56,20 +60,23 @@ const SortableChip = SortableElement((props: SortableChipProps) => {
...
             onClick={handleClose}
             data-test-id="button-close"
+            disabled={loading}
           >
-            <CloseIcon className={classes.closeIcon} />
+            <CloseIcon
+              className={clsx(classes.closeIcon, loading && classes.disabled)}
```
Comment (poulch) [inline src/components/SortableChip/SortableChip.tsx:78]:
```suggestion
     className={clsx(classes.sortableHandle, {
              [classes.disabled]: loading
          )}
```

**Review by poulch — APPROVED** [review]

**Review by poulch — APPROVED** [review]

**Review by wojteknowacki — APPROVED** [review]

**Pull request merged**